doc/WAL.md

```diff
@@ -0,0 +1,39 @@
+
+# WAL Reader
+
+This document provides an overview of the `wal_reader` tool, particularly focusing on the `parse_wal_file` function, which is the main entry point for parsing Write-Ahead Log (WAL) files. Right now this function only parses the given WAL file and prints the description of each record but in future, it will be integrated with the other parts of the code.
+
```

Comment:
Can you explain the different WAL format from PostgreSQL 13 to PostgreSQL 17 ?

Is there an easy place to see that ?

Reply:
The only difference I could notice was in the `pgmoneta_wal_is_bkp_image_compressed` function. Just added the explanation to this part.

Reply:
Well, I mean, the different PostgreSQL have different fields depending on the version...

* PostgreSQL 13: 0xD106
* PostgreSQL 14: 0xD10D
* PostgreSQL 15: 0xD110
* PostgreSQL 16: 0xD113
* PostgreSQL 17: 0xD116 (currently)

You can see commits like https://github.com/postgres/postgres/commit/2b5819e2b4d6c69676c61ee799fd9590be2308ce that will bump the WAL identifier

Reply:
So, I would like to see a description of how those different formats are supported through the https://github.com/pgmoneta/pgmoneta/blob/main/src/include/pgmoneta.h#L226 field

Reply:
Okay, I see. The magic values you mentioned are currently stored in the `xlp_magic` in the `xlog_page_header_data` struct, but I don't do any processing based on them and they're just stored in the struct.
The code was able to parse the sample WAL files I generated for testing on Postgresql 13 -> 17, but I would also double-check the commits you sent to note any other fields that may not be covered in my tests

Reply:
As you can see `wal_level` was added to `struct xl_end_of_recovery` and `struct CheckPoint` - that is available in 0xD116, but not in earlier versions

Reply:
So, you should explain how you handle the parsing and the different version since the structs will have different fields

Reply:
I'm thinking about using a wrapper struct with a factory pattern. We can implement it for `xl_end_of_recovery` and use the same approach anytime a change is made in one of the structs. What do you think about this?
```
struct xl_end_of_recovery_v16 {
    timestamp_tz end_time;
    timeline_id this_timeline_id;
    timeline_id prev_timeline_id;
};

struct xl_end_of_recovery_v17 {
    timestamp_tz end_time;
    timeline_id this_timeline_id;
    timeline_id prev_timeline_id;
    int wal_level;
};

struct xl_end_of_recovery {
    int pg_version;
    union {
        struct xl_end_of_recovery_v16 v16;
        struct xl_end_of_recovery_v17 v17;
    } data;
    void (*parse)(struct xl_end_of_recovery* wrapper, const void* rec);
    char* (*format)(struct xl_end_of_recovery* wrapper, char* buf);
};
xl_end_of_recovery* create_xl_end_of_recovery(int pg_version) {
    xl_end_of_recovery* wrapper = malloc(sizeof(xl_end_of_recovery));
    wrapper->pg_version = pg_version;
    
    if (pg_version >= 17) {
        wrapper->parse = parse_v17;
        wrapper->format = format_v17;
    } else {
        wrapper->parse = parse_v16;
        wrapper->format = format_v16;
    }
    
    return wrapper;
}

void parse_v16(xl_end_of_recovery* wrapper, const void* rec) {
    memcpy(&wrapper->data.v16, rec, sizeof(struct xl_end_of_recovery_v16));
}

void parse_v17(xl_end_of_recovery* wrapper, const void* rec) {
    memcpy(&wrapper->data.v17, rec, sizeof(struct xl_end_of_recovery_v17));
}

char* format_v16(xl_end_of_recovery* wrapper, char* buf) {
    struct xl_end_of_recovery_v16* xlrec = &wrapper->data.v16;
    return pgmoneta_format_and_append(buf, "tli %u; prev tli %u; time %s",
                                      xlrec->this_timeline_id, xlrec->prev_timeline_id,
                                      pgmoneta_wal_timestamptz_to_str(xlrec->end_time));
}

char* format_v17(xl_end_of_recovery* wrapper, char* buf) {
    struct xl_end_of_recovery_v17* xlrec = &wrapper->data.v17;
    return pgmoneta_format_and_append(buf, "tli %u; prev tli %u; time %s; wal_level %d",
                                      xlrec->this_timeline_id, xlrec->prev_timeline_id,
                                      pgmoneta_wal_timestamptz_to_str(xlrec->end_time),
                                      xlrec->wal_level);
}
```

Reply:
Yeah, that could be a solution.

Do you have a list of changes between the different versions ? Do we need a lot of these parse/format functions ?

Reply:
Unfortunately, we have to go through the history of `src/include/access/xlog_internal.h` and investigate the commits that bumped the magic number.

There is no official list of changes :disappointed:

Reply:
Yeah I guess that would take a lot of time to check all structs manually :(

Reply:
Yeah, it isn't ideal.

But, commits like https://github.com/postgres/postgres/commit/f83d709760d84253b539ee454ce9ce1ceb9ad9ac changed existing `struct`s and added new ones

Reply:
Just to make it clear -  we only need to support the official magic numbers.

So, you need to find those commits, and then diff between them in one go

Reply:
E.g. 0xD106 -> 0xD10D, 0xD10D -> 0xD110, and so on

Reply:
Ok I see, so in total there should be like 4-5 commits that should be diffed, this is not hard and can be handled

Reply:
You can document the actual commit ids for each magic number change in the developer guide if you want.

Some of the commits are just year bumps, and so on

Reply:
E.g.

```
0xD116: 2b5819e2b4d6c69676c61ee799fd9590be2308ce
...
```

Reply:
This is the list of commits for official changes.
I'm looking at the differences between version 17 and 16 and it's huge! ~4000 files changed with ~3500 of them in the src directory and ~700 in the `src/include`, so manual checking of every change in every file is impossible. I'm going to only check file changes in `src/include/access` files that are ~90 file changes for the last commit. Does the strategy sound good to you? or maybe you have better suggestions to narrow it down?
```
PostgreSQL 13: 0xD106: c6b92041d38512a4176ed76ad06f713d2e6c01a8
PostgreSQL 14: 0xD10D: 08aa89b326261b669648df97d4f2a6edba22d26a
PostgreSQL 15: 0xD110: 8b1dccd37c71ed2ff016294d8f9053a32b02b19e
PostgreSQL 16: 0xD113: 6af1793954e8c5e753af83c3edb37ed3267dd179
PostgreSQL 17: 0xD116: 402b586d0a9caae9412d25fcf1b91dae45375833
```

Reply:
Yes, it should be ok to just verify the `src/include/access/` directory, so

```
git diff 6af1793954e8c5e753af83c3edb37ed3267dd179..402b586d0a9caae9412d25fcf1b91dae45375833 src/include/access/
```

Reply:
.. and you can use a lot of `| grep -v`s to get it down even further

Reply:
I have gone through the commits version by version and compiled a list of changes here: https://elastic-traffic-761.notion.site/Postgres-WAL-diff-5d7f873c9d204ff79cb6779262c76326.

I tried my best not to miss any changes that are related to Pgmoneta, but since the process an overlook is possible. Can you please review it and let me know if it's good to go? I can start implementing the solution above for the documented structs.

Reply:
Yeah, it looks good - you can put this information into the developer guide